README.md
# Razer Drivers

**This repository contains the legacy razer drivers source based on the [razer_chroma_drivers](https://github.com/pez2001/razer_chroma_drivers) project. Due to code quality issues a complete rewrite can be found [here](https://github.com/openrazer/razer-drivers)**

This repository holds the source code for the linux kernel drivers to support various Razer devices.
All notable changes to this project will be documented in the **[CHANGELOG.md](CHANGELOG.md)** file.

## Supported Devices

| NAME                                       | TYPE     | DRIVER SUPPORT                             |
|:-------------------------------------------|:---------|:-------------------------------------------|
| Razer Firefly                              | Mousemat | all lighting modes                         |
| Razer Mamba                                | Mouse    | all lighting modes                         |
| Razer Abyssus                              | Mouse    | all modes bar the refresh rate             |
| Razer Blackwidow Classic                   | Keyboard | all lighting modes                         |
| Razer Blackwidow Ultimate 2012             | Keyboard | all lighting modes                         |
| Razer Blackwidow Ultimate 2013             | Keyboard | all lighting modes                         |
| Razer Blackwidow Chroma                    | Keyboard | all lighting modes                         |
| Razer Blackwidow Chroma Tournament Edition | Keyboard | all lighting modes                         |
| Razer Blackwidow Ultimate 2016             | Keyboard | all bar custom lighting                    |
| Razer Blade Stealth                        | Keyboard | all lighting modes                         |
| Razer Blade 14 2016                        | Keyboard | partially all lighting modes - In Progress |


## Installation

### Automatic installation with DKMS

This requires DKMS to be installed.

```
    make DESTDIR=/ install_dkms
```

### Manual installation

This requires recompilation on every kernel update.

```
    make all
    make DESTDIR=/ install
```

## Authors

Special thanks to Tim Theede, the initial creator of the [razer_chroma_drivers project](https://github.com/pez2001/razer_chroma_drivers) on which this project is based on.

A list of all authors contributed to this project can be found in the **[AUTHORS.md](AUTHORS.md)** file.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This project follows the [Semantic Versioning](http://semver.org/).

## v0.0.1 - 2016-07-15
- Initial first release. Mainly based on the previous razer_chroma_drivers with additional initial support for the Razer Blade 14 2016.


src/razercommon.c
/*
 * Razer Kernel Drivers
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 */

#include <linux/kernel.h>
#include <linux/slab.h>
#include <linux/module.h>
#include <linux/init.h>
#include <linux/usb/input.h>
#include <linux/hid.h>

#include "razercommon.h"

/**
 * Send USB control report to the keyboard
 * USUALLY index = 0x02
 * FIREFLY is 0
 */
int razer_send_control_msg(struct usb_device *usb_dev, void const *data, uint report_index, ulong wait_min, ulong wait_max)
{
	uint request = HID_REQ_SET_REPORT; // 0x09
	uint request_type = USB_TYPE_CLASS | USB_RECIP_INTERFACE | USB_DIR_OUT; // 0x21
    uint value = 0x300;
    uint size = RAZER_USB_REPORT_LEN;
    char *buf;
    int len;

    buf = kmemdup(data, size, GFP_KERNEL);
    if (buf == NULL)
    {
        return -ENOMEM;
    }

    // Send usb control message
    len = usb_control_msg(usb_dev, usb_sndctrlpipe(usb_dev, 0),
            request,      // Request
            request_type, // RequestType
            value,        // Value
            report_index, // Index
            buf,          // Data
            size,         // Length
            USB_CTRL_SET_TIMEOUT);

    // Wait
    usleep_range(wait_min, wait_max);

    // Free the memory again.
    kfree(buf);

    if(len!=size)
    {
        printk(KERN_WARNING "razer driver: Device data transfer failed.");
    }

    return ((len < 0) ? len : ((len != size) ? -EIO : 0));
}

/**
 * Get a response from the razer device
 *
 * Makes a request like normal, this must change a variable in the device as then we
 * tell it give us data and it gives us a report.
 *
 * Supported Devices:
 *   Razer Chroma
 *   Razer Mamba
 *   Razer BlackWidow Ultimate 2013*
 *   Razer Firefly*
 *
 * Request report is the report sent to the device specifing what response we want
 * Response report will get populated with a response
 *
 * Returns 0 when successful, 1 if the report length is invalid.
 */
int razer_get_usb_response(struct usb_device *usb_dev, uint report_index, struct razer_report* request_report, uint response_index, struct razer_report* response_report, ulong wait_min, ulong wait_max)
{
    uint request = HID_REQ_GET_REPORT; // 0x01
    uint request_type = USB_TYPE_CLASS | USB_RECIP_INTERFACE | USB_DIR_IN; // 0xA1
    uint value = 0x300;

    uint size = RAZER_USB_REPORT_LEN; // 0x90
    int len;
    int retval;
    int result = 0;

    memset(response_report, 0, sizeof(struct razer_report));

    // Send the request to the device.
    // TODO look to see if index needs to be different for the request and the response
    retval = razer_send_control_msg(usb_dev, request_report, report_index, wait_min, wait_max);

    // Now ask for reponse
    len = usb_control_msg(usb_dev, usb_rcvctrlpipe(usb_dev, 0),
          request,         // Request
          request_type,    // RequestType
          value,           // Value
          response_index,  // Index
          response_report, // Data
          size,
          USB_CTRL_SET_TIMEOUT);

    usleep_range(wait_min, wait_max);

    // Error if report is wrong length
    if(len != RAZER_USB_REPORT_LEN)
    {
        printk(KERN_WARNING "razer driver: Invalid USB repsonse. USB Report length: %d\n", len);
        result = 1;
    }

    return result;
}

/**
 * Calculate the checksum for the usb message
 *
 * Checksum byte is stored in the 2nd last byte in the messages payload.
 * The checksum is generated by XORing all the bytes in the report starting
 * at byte number 2 (0 based) and ending at byte 88.
 */
unsigned char razer_calculate_crc(struct razer_report *report)
{
    /*second to last byte of report is a simple checksum*/
    /*just xor all bytes up with overflow and you are done*/
    unsigned char crc = 0;
    unsigned char *_report = (unsigned char*)report;

    unsigned int i;
    for(i = 2; i < 88; i++)
    {
        crc ^= _report[i];
    }

    return crc;
}

/**
 * Get initialised razer report
 */
struct razer_report get_razer_report(unsigned char command_class, unsigned char command_id, unsigned char data_size)
{
    struct razer_report new_report;
    memset(&new_report, 0, sizeof(struct razer_report));

    new_report.status = 0x00;
    new_report.transaction_id.id = 0xFF;
    new_report.remaining_packets = 0x00;
    new_report.protocol_type = 0x00;
    new_report.command_class = command_class;
    new_report.command_id.id = command_id;
    new_report.data_size = data_size;

    return new_report;
}

/**
 * Get empty razer report
 */
struct razer_report get_empty_razer_report(void)
{
    struct razer_report new_report;
    memset(&new_report, 0, sizeof(struct razer_report));

    return new_report;
}

/**
 * Print report to syslog
 */
void print_erroneous_report(struct razer_report* report, char* driver_name, char* message)
{
    printk(KERN_WARNING "%s: %s. Start Marker: %02x id: %02x Num Params: %02x Reserved: %02x Command: %02x Params: %02x%02x%02x%02x%02x%02x .\n",
        driver_name,
        message,
        report->status,
        report->transaction_id.id,
        report->data_size,
        report->command_class,
        report->command_id.id,
        report->arguments[0], report->arguments[1], report->arguments[2], report->arguments[3], report->arguments[4], report->arguments[5]);
}


src/razercommon.h
/*
 * Razer Kernel Drivers
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 */

#ifndef DRIVER_RAZERCOMMON_H_
#define DRIVER_RAZERCOMMON_H_


//#################//
//### Constants ###//
//#################//

// The Razer vendor ID
#define USB_VENDOR_ID_RAZER 0x1532

// Each USB report has 90 bytes
#define RAZER_USB_REPORT_LEN 0x5A



//#############//
//### Types ###//
//#############//

struct razer_report;

struct razer_rgb {
    unsigned char r,g,b;
};

union transaction_id_union {
    unsigned char id;
    struct transaction_parts {
        unsigned char device : 3;
        unsigned char id : 5;
    } parts;
};

union command_id_union {
    unsigned char id;
    struct command_id_parts {
        unsigned char direction : 1;
        unsigned char id : 7;
    } parts;
};

/* Status:
 * 0x00 New Command
 * 0x01 Command Busy
 * 0x02 Command Successful
 * 0x03 Command Failure
 * 0x04 Command No Response / Command Timeout
 * 0x05 Command Not Support
 *
 * Transaction ID used to group request-response, device useful when multiple devices are on one usb
 * Remaining Packets is the number of remaining packets in the sequence
 * Protocol Type is always 0x00
 * Data Size is the size of payload, cannot be greater than 80. 90 = header (8B) + data + CRC (1B) + Reserved (1B)
 * Command Class is the type of command being issued
 * Command ID is the type of command being send. Direction 0 is Host->Device, Direction 1 is Device->Host. AKA Get LED 0x80, Set LED 0x00
 *
 * */

struct razer_report {
    unsigned char status;
    union transaction_id_union transaction_id; /* */
    unsigned short remaining_packets; /* Big Endian */
    unsigned char protocol_type; /*0x0*/
    unsigned char data_size;
    unsigned char command_class;
    union command_id_union command_id;
    unsigned char arguments[80];
    unsigned char crc;/*xor'ed bytes of report*/
    unsigned char reserved; /*0x0*/
};



//#################//
//### Functions ###//
//#################//

int razer_send_control_msg(struct usb_device *usb_dev,void const *data, uint report_index, ulong wait_min, ulong wait_max);
int razer_get_usb_response(struct usb_device *usb_dev, uint report_index, struct razer_report* request_report, uint response_index, struct razer_report* response_report, ulong wait_min, ulong wait_max);
unsigned char razer_calculate_crc(struct razer_report *report);
struct razer_report get_razer_report(unsigned char command_class, unsigned char command_id, unsigned char data_size);
struct razer_report get_empty_razer_report(void);
void print_erroneous_report(struct razer_report* report, char* driver_name, char* message);

#endif /* DRIVER_RAZERCOMMON_H_ */


src/razerkbd_driver.h
/*
 * Razer Kernel Drivers
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 */

#ifndef __HID_RAZER_KEYBOARD_BLACKWIDOW_CHROMA_H
#define __HID_RAZER_KEYBOARD_BLACKWIDOW_CHROMA_H


//#################//
//### Constants ###//
//#################//

#define RAZER_KBD_WAIT_MIN_US 600
#define RAZER_KBD_WAIT_MAX_US 800

// Device IDs
#define USB_DEVICE_ID_RAZER_BLACKWIDOW_ORIGINAL 0x011B
#define USB_DEVICE_ID_RAZER_BLACKWIDOW_ULTIMATE_2012 0x010D
#define USB_DEVICE_ID_RAZER_BLACKWIDOW_ULTIMATE_2013 0x011A
#define USB_DEVICE_ID_RAZER_BLACKWIDOW_ULTIMATE_2016 0x0214
#define USB_DEVICE_ID_RAZER_BLADE_STEALTH 0x0205
#define USB_DEVICE_ID_RAZER_BLADE_14_2016 0x020F
#define USB_DEVICE_ID_RAZER_BLACKWIDOW_CHROMA 0x0203
#define USB_DEVICE_ID_RAZER_BLACKWIDOW_CHROMA_TE 0x0209

#define RAZER_BLACKWIDOW_CHROMA_ROW_LEN 0x16
#define RAZER_STEALTH_ROW_LEN 0x10
#define RAZER_BLADE_14_2016_ROW_LEN 0x10



//#############//
//### Types ###//
//#############//

struct razer_kbd_device {
    struct usb_device *usbdev;
    struct hid_device *hiddev;
    unsigned char effect;
    char name[128];
    char phys[64];
    bool effect_submitted;
};

#endif


src/razerfirefly_driver.h
/*
 * Razer Kernel Drivers
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 */

#ifndef __HID_RAZER_FIREFLY_H
#define __HID_RAZER_FIREFLY_H


//#################//
//### Constants ###//
//#################//

#define USB_DEVICE_ID_RAZER_FIREFLY 0x0C00

#define RAZER_FIREFLY_ROW_LEN 0x0F

#define RAZER_FIREFLY_WAIT_MIN_US 900
#define RAZER_FIREFLY_WAIT_MAX_US 1000



//#############//
//### Types ###//
//#############//

struct razer_firefly_device {
    struct usb_device *usbdev;
    struct hid_device *hiddev;
    unsigned char effect;
    char name[128];
    char phys[64];
    bool effect_submitted;
    int brightness;
};


#endif


src/razermouse_driver.h
/*
 * Razer Kernel Drivers
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 */

#ifndef __HID_RAZER_MOUSE_H
#define __HID_RAZER_MOUSE_H


//#################//
//### Constants ###//
//#################//

#define USB_DEVICE_ID_RAZER_MAMBA 0x0045
#define USB_DEVICE_ID_RAZER_ABYSSUS 0x0042

#define RAZER_MAMBA_ROW_LEN 15

#define RAZER_MOUSE_WAIT_MIN_US 600
#define RAZER_MOUSE_WAIT_MAX_US 800



//#############//
//### Types ###//
//#############//

struct razer_mouse_device {
    struct usb_device *usbdev;
    struct hid_device *hiddev;
    unsigned char effect;
    char name[128];
    char phys[64];
};

#endif
